zerver/lib/actions.py

```diff
@@ -911,6 +911,9 @@ def do_deactivate_user(user_profile: UserProfile,
             do_deactivate_user(profile, acting_user=acting_user, _cascade=False)
 
 def do_deactivate_stream(stream: Stream, log: bool=True, acting_user: Optional[UserProfile]=None) -> None:
+    # We want to mark all messages in the stream as read.
+    UserMessage.objects.filter(message__recipient_id=stream.recipient_id) \
+        .extra(where=[UserMessage.where_unread()]).update(flags=F('flags').bitor(UserMessage.flags.read))
```

Comment:
Hmm, this query could be very expensive.  I think we might want to use the `deferred_work` queue processor for this, batch the query with 100 messages to the stream at a time.

Otherwise, if the stream had 1000 subscribers who haven't read 10000 messages, this query could theoretically trigger a single database transaction that needs to make 10M writes, which won't succeed.

Reply:
Moved to deferred_Work and spit into batches now.

Reply:
This should be safe enough; given the testing above, an individual query covering all messages in a large stream could run for 100s or more, which is a longer transaction than we ever want to have running.